src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1277,28 +1277,15 @@ public Boolean visitErrorType(ErrorType t, Type s) {
         public NullabilityComparator nullabilityComparator = new NullabilityComparator();
         public class NullabilityComparator extends TypeRelation {
             BiFunction<Type, Type, Boolean> differentNullability;
-            Warner warner;
 
             NullabilityComparator reset(BiFunction<Type, Type, Boolean> differentNullability) {
                 this.differentNullability = differentNullability;
-                if (this.warner == null || this.warner == noWarnings) {
-                    this.warner = !warnStack.isEmpty() ? warnStack.head : noWarnings;
-                }
-                return this;
-            }
-
-            public NullabilityComparator setWarner(Warner warner) {
-                this.warner = warner;
-                return this;
-            }
-
-            public NullabilityComparator clearWarner() {
-                this.warner = null;
                 return this;
             }
 
             @Override
             public Boolean visitType(Type t, Type s) {
+                Warner warner = !warnStack.isEmpty() ? warnStack.head : noWarnings;
```

Comment:
Good cleanup. I assume that now nullability comparator really has no state, so you could in principle create two instances (with the two comparator functions), and just use those instances, right?

Reply:
not really, in one case we are using this comparator: `(t1, t2) -> !hasSameNullability(t1, t2)` so we would need three different instances

Reply:
I see - but still - there is no state in the comparator - right? E.g. in principle clients can allocate a new one and take it from there each time. So perhaps that would be more honest, rather than having a shared instance but then having to call a `reset` method?